### PR TITLE
[torch] remove deprecated uint8 in favor of bool

### LINKED ISSUE #21013: HF models use deprecated pytorch function invocations
Masks are frequently created in `uint8` type, see e.g. here https://github.com/huggingface/transformers/blob/8fb4d0e4b46282d96386c229b9fb18bf7c80c25a/src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py#L186 or here https://github.com/huggingface/transformers/blame/52dd2b61bff8af5b6409fdd5ec92a9b3114f3636/src/transformers/models/codegen/modeling_codegen.py#L101, and then used in `torch.where`. Use of `uint8` masks in `torch.where` has been deprecated for couple years, and though it still works in pytorch eager (with a warning), support for this has been removed in `torch.compile`. It would be good to audit places where uint8 masks are used and replace them with bool masks.

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
This should fix #21013. 
Have not ran the tests yet, so leaving as draft

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Waiting for the CI to work again.

**Comment by ArthurZucker** [comment]
Ok now the real clean up is starting, have to identify all attention masks, vs the causal masks. I think it is a good thing, will help understanding

**Comment by ArthurZucker** [comment]
To Do;

- [ ] Fix all usages of `attention_mask` where operation are done assuming that the tensor is not made of `bool`. 
- [ ] Add docstring to specify that the attention masks should be a bool tensor everywhere
- [ ] Update the type hints for all of the forward passes

**Comment by ArthurZucker** [comment]
Last check, I need to make sure these new mask don't go through a `1.0 - mask` afterwards and will be good to go

File: src/transformers/models/longformer/modeling_longformer.py [hunk 1]
```diff
@@ -404,12 +404,16 @@ def _compute_global_attention_mask(input_ids, sep_token_id, before_sep_token=Tru
     # bool attention mask with True in locations of global attention
     attention_mask = torch.arange(input_ids.shape[1], device=input_ids.device)
     if before_sep_token is True:
-        attention_mask = (attention_mask.expand_as(input_ids) < question_end_index).to(torch.uint8)
+        attention_mask = (attention_mask.expand_as(input_ids) < question_end_index).to(
+            torch.bool
+        )  # previously `dtype=torch.uint8`, cf pytorch 1.2.0 compatibility
```
Comment (sgugger) [inline src/transformers/models/longformer/modeling_longformer.py:409]: I don't think we need the comment, but if we keep it, let's put it above the line of code as it screws up formatting.

File: src/transformers/models/longformer/modeling_longformer.py [hunk 2]
```diff
@@ -404,12 +404,16 @@ def _compute_global_attention_mask(input_ids, sep_token_id, before_sep_token=Tru
...
     else:
         # last token is separation token and should not be counted and in the middle are two separation tokens
-        attention_mask = (attention_mask.expand_as(input_ids) > (question_end_index + 1)).to(torch.uint8) * (
+        attention_mask = (attention_mask.expand_as(input_ids) > (question_end_index + 1)).to(torch.bool) * (
             attention_mask.expand_as(input_ids) < input_ids.shape[-1]
-        ).to(torch.uint8)
+        ).to(
+            torch.bool
+        )  # previously `dtype=torch.uint8`, cf pytorch 1.2.0 compatibility
```
Comment (sgugger) [inline src/transformers/models/longformer/modeling_longformer.py:416]: Same here.

File: src/transformers/models/reformer/modeling_reformer.py [hunk 3]
```diff
@@ -666,7 +666,9 @@ def _hash_vectors(self, vectors, num_hashes, attention_mask, increase_num_bucket
             # add an extra bucket for padding tokens only
             num_buckets = num_buckets + 1
             # assign padding tokens extra bucket
-            buckets_mask = attention_mask.to(torch.uint8)[:, None, None, :].expand(buckets.shape)
+            buckets_mask = attention_mask.to(torch.bool)[:, None, None, :].expand(
+                buckets.shape
+            )  # previously `dtype=torch.uint8`, cf pytorch 1.2.0 compatibility
```
Comment (sgugger) [inline src/transformers/models/reformer/modeling_reformer.py:671]: Same here.

File: src/transformers/models/reformer/modeling_reformer.py [hunk 4]
```diff
@@ -841,7 +843,9 @@ def _compute_attn_mask(
         # attention mask for LSH
         if attention_mask is not None:
             # if chunked attention, the attention mask has to correspond to LSH order
-            attention_mask = attention_mask.to(torch.uint8)[:, None, :]
+            attention_mask = attention_mask.to(torch.bool)[
+                :, None, :
+            ]  # previously `dtype=torch.uint8`, cf pytorch 1.2.0 compatibility
```
Comment (sgugger) [inline src/transformers/models/reformer/modeling_reformer.py:848]: Same there.

File: src/transformers/models/reformer/modeling_reformer.py [hunk 5]
```diff
@@ -1225,7 +1229,9 @@ def _compute_attn_mask(
     ):
         # chunk attention mask and look before and after
         if attention_mask is not None:
-            attention_mask = attention_mask.to(torch.uint8)[:, None, :]
+            attention_mask = attention_mask.to(torch.bool)[
+                :, None, :
+            ]  # previously `dtype=torch.uint8`, cf pytorch 1.2.0 compatibility
```
Comment (sgugger) [inline src/transformers/models/reformer/modeling_reformer.py:1234]: Same there.

**Review by sgugger — APPROVED** [review]
LGTM, thanks!

**Comment by ArthurZucker** [comment]
Test failing is uneralted, merging
